api/api.go

```diff
@@ -114,3 +126,40 @@ func NewAPI(config *conf.Configuration, db *gorm.DB, mailer *mailer.Mailer) *API
 	api.handler = corsHandler.Handler(mux)
 	return api
 }
+
+func (a *API) trace(f func(context.Context, http.ResponseWriter, *http.Request)) func(ctx context.Context, w http.ResponseWriter, r *http.Request) {
+	return func(ctx context.Context, w http.ResponseWriter, r *http.Request) {
+		id := uuid.NewRandom().String()
+		log := a.log.WithField("request_id", id)
+
+		log = log.WithFields(logrus.Fields{
+			"method": r.Method,
+			"path":   r.URL.Path,
+		})
+
+		// optionally add the user id stuff
+		claims := getClaims(ctx)
+		if claims != nil {
+			log = log.WithFields(logrus.Fields{
+				"claim_id":     claims.ID,
+				"claim_groups": claims.Groups,
+				"claim_email":  claims.Email,
+			})
+		}
+
+		ctx = withRequestID(ctx, id)
+		ctx = withLogger(ctx, log)
+		ctx = withConfig(ctx, a.config)
+
+		rp := ResponseProxy{ResponseWriter: w}
+		startTime := time.Now()
+
+		log.Info("request started")
+		defer log.WithFields(logrus.Fields{
+			"status":   rp.statusCode,
```

Comment:
Pros of this trace function:

- It doesn't send the LogEntry in the context. (As you'll see later, this is not actually a pro)

Cons of this trace function:

- It needs to implement all the ResponseProxy stuff, copying Kami's implementation.
- It needs to wrap every single endpoint with this function.
- It introduces a bug in the defer function.
- It doesn't allow to add more information to the log entry, which would improve even more request tracing.

Given all this, I think we should use Kami's middlewares rather than implementing our own.